lib/bundler/resolver.rb

```diff
@@ -146,7 +146,25 @@ def search_for(dependency)
           @gem_version_promoter.sort_versions(dependency, spec_groups)
         end
       end
-      search.select {|sg| sg.for?(platform) }.each {|sg| sg.activate_platform!(platform) }
+      selected_sgs = []
+      search.each do |sg|
+        spec = sg.spec_for(platform)
+        next if spec.nil?
+        # Add a spec group for "non platform specific spec" as the fallback
+        # spec group.
+        sg_ruby = sg.copy_for(Gem::Platform::RUBY)
+        selected_sgs << sg_ruby if sg_ruby
+        sg_all_platforms = nil
+        self.class.sort_platforms(@platforms).reverse_each do |other_platform|
+          if sg_all_platforms.nil?
+            sg_all_platforms = sg.copy_for(other_platform)
+          else
+            sg_all_platforms.activate_platform!(other_platform)
+          end
+        end
+        selected_sgs << sg_all_platforms if sg_all_platforms
```

Comment:
`if sg_all_platforms` may not be needed.

Reply:
I don't think it's needed either. At least I think `@platforms` should never be empty here (which would cause this to crash), because in that case we would abort earlier (I think [here](https://github.com/rubygems/bundler/blob/4cc9dc992b2fa38c3cd157d5d1eb2396bac57a47/lib/bundler/definition.rb#L499-L509)?). Not source about the `copy_for` inside the loop always returning `nil` though.

I think we could maybe add a note that we could try to remove that clause in the future, and do it in the future once all these changes stabilize and we feel everything regarding multiplatform is solid.

Reply:
OK. I've removed the `if` by adding the `platform` to target platforms. The `platform` must be valid because we check it by `sg.for?(platform)`. It means that at least one `sg.copy_for(other_platform)` succeeds. So we can remove the `if` safely.

Reply:
Nice, that makes sense!